embrace-android-sdk/src/main/java/io/embrace/android/embracesdk/capture/crash/CrashDataSourceImpl.kt

```diff
@@ -0,0 +1,131 @@
+     *
+     * @param thread    the crashing thread
+     * @param exception the exception thrown by the thread
+     */
+    override fun handleCrash(thread: Thread, exception: Throwable) {
+        if (!mainCrashHandled) {
+            mainCrashHandled = true
+
+            // Stop ANR tracking first to avoid capture ANR when crash message is being sent
+            anrService?.forceAnrTrackingStopOnCrash()
+
+            // Check if the unity crash id exists. If so, means that the native crash capture
+            // is enabled for an Unity build. When a native crash occurs and the NDK sends an
+            // uncaught exception the SDK assign the unity crash id as the java crash id.
+            val crashId = ndkService.getUnityCrashId() ?: getEmbUuid()
+            val crashNumber = preferencesService.incrementAndGetCrashNumber()
+            val crashAttributes = TelemetryAttributes(
+                sessionProperties = sessionProperties
+            )
+
+            // TODO: make sure serialized JSON values in strings is corrected and as expected
+            val crashException = LegacyExceptionInfo.ofThrowable(exception)
+            crashAttributes.setAttribute(exceptionType, crashException.name)
+            crashAttributes.setAttribute(exceptionMessage, crashException.message ?: "")
+            crashAttributes.setAttribute(exceptionStacktrace, crashException.lines.toString())
+            crashAttributes.setAttribute(logRecordUid, crashId)
+            crashAttributes.setAttribute(embCrashNumber, crashNumber.toString())
+            crashAttributes.setAttribute(EmbType.System.Crash.embAndroidCrashExceptionCause, getCause(exception.cause))
+            crashAttributes.setAttribute(embAndroidThreads, crashNumber.toString())
+
+            // TODO: add threading through of jsException
+            jsException?.let { crashAttributes.setAttribute(embAndroidReactNativeCrashJsExceptions, jsException.toString()) }
+
+            val logEventData = LogEventData(
+                schemaType = getSchemaType(crashAttributes),
+                message = "",
+                severity = Severity.ERROR,
+            )
+
+            logWriter.addLog(logEventData)
+
+            // TODO: apply gating service logic
+
+            // Attempt to end and send the session
+            sessionOrchestrator.endSessionWithCrash(crashId)
+
+            // Indicate that a crash happened so we can know that in the next launch
+            crashMarker.mark()
+        }
+    }
+
+    override fun logUnhandledJsException(exception: JsException) {
+        this.jsException = exception
+    }
+
+    private fun getSchemaType(attributes: TelemetryAttributes): SchemaType =
+        if (attributes.getAttribute(embAndroidReactNativeCrashJsExceptions) != null) {
+            SchemaType.ReactNativeCrash(attributes)
+        } else {
+            SchemaType.Crash(attributes)
+        }
+
+    private fun getCause(throwable: Throwable?): String {
```

Comment:
This might iterate infinitely if an exception in the `cause` chain is the same object as the root throwable. This sounds unlikely but it will happen in production! To handle this scenario I'd keep a collection of all the previously added throwables & return early from the function if we spot one we've already seen

Reply:
Thanks for the feedback Jamie. I'll attack this in a separate PR as the intention of this one is just to add the ability to switch between both services. This was a previous code that was moved and still need iteration.

Reply:
Updated.